### DISCUSSION TITLE
error setting cgroup config for procHooks process: unable to freeze: unknown

### DISCUSSION THREAD

**Discussion by dabcoder** [opencontainers/runc#4232] [General]
Hello,

I am looking for insights for this particular error for a container that tried to start:

```
failed to create containerd task: failed to create shim task: OCI
            runtime create failed: runc create failed: unable to start container
            process: error during container init: error setting cgroup config
            for procHooks process: unable to freeze: unknown 
```

I've located that error [here](https://github.com/opencontainers/runc/blob/8e69225e2e5f25c46e79cf49935db362ead8f496/libcontainer/cgroups/fs/freezer.go#L101) unless I'm mistaken. A couple of questions: 

- as far as I understand, this is specific to an OS that would use `cgroups` v1 only, not v2. Is that correct? 
(I also had a look at https://docs.kernel.org/admin-guide/cgroup-v1/freezer-subsystem.html). 
- How does that process work for cgroups v2 which doesn't have that subsystem (starting or stopping required tasks)?

Thanks in advance.

**Comment by cyphar** [comment]
> How does that process work for cgroups v2 which doesn't have that subsystem (starting or stopping required tasks)?

cgroupv2 does have freezing, it's just implemented as part of the cgroup core rather than as a separate controller. See `/sys/fs/cgroup/.../cgroup.freeze`. The design of the cgroupv2 freezer is far more robust than cgroupv1, which is why the bug can only occur on cgroupv1.

**Reply by dabcoder** [reply]
Thanks @cyphar, all clear. Closing this discussion.

**Reply by dabcoder** [reply]
Forgot to ask, what does the last `unknown` error refer to (after `unable to freeze:`)? Wasn't able to locate it in the codebase.

**Reply by cyphar** [reply]
It probably comes from containerd. [A quick look at the codebase indicates it might refer to the container state after the error, but I don't know.](https://github.com/containerd/containerd/blob/92900bf7300b4872db2a43e82128fda9cde08a79/client/task.go#L96)
